addons/crm/models/crm_lead.py

```diff
@@ -1377,10 +1377,15 @@ def _pls_update_frequency_table(self, values_to_create, stage_ids, stage_sequenc
             return values_to_create
 
         fields = ['stage_id', 'team_id'] + self._pls_get_safe_fields()
-        frequencies = dict((field, {}) for field in (fields + ['tag_id']))
+        tag_ids = []
+        if 'tag_ids' in fields:
+            fields.remove('tag_ids')
+            tag_ids = ['tag_id']
+        frequencies = dict((field, {}) for field in (fields + tag_ids))
 
         frequencies = self._pls_update_frequency_table_fields(frequencies, stage_ids, stage_sequences, fields, team_id, pls_start_date)
-        frequencies = self._pls_update_frequency_table_tag(frequencies, team_id, pls_start_date)
+        if tag_ids:
+            frequencies = self._pls_update_frequency_table_tag(frequencies, team_id, pls_start_date)
```

Comment:
you can keep `frequencies = dict((field, {}) for field in fields)` -> line 1380 to 1384 are not necessary
and check `if 'tag_ids' in fields` to set frequency
include _pls_update_frequency_table_tag into _pls_update_frequency_table_fields and call it only if 'tag_ids' in fields`

Reply:
done, I moved _pls_update_frequency_table_tag inside _pls_update_frequency_table_fields, but I think that I still need to remove tag_ids from fields because when we will enter in _pls_update_frequency_table_fields, the query will crash because there is no column tag_ids in the db.

Reply:
Yes but you can do this inside _pls_update_frequency_table_fields.
1. extract tag_ids field and set use_tag variable
2. do what the _pls_update_frequency_table_fields content already did
3. if use_tag, call  _pls_update_frequency_table_tag
4. return
All this inside _pls_update_frequency_table_fields.

Reply:
done

Reply:
Well, I did not understand/remember directly that we had to change 'tag_ids' into 'tag_id'.
What I'll do then is this, to put this logic at the same place :
```py
fields = ['stage_id', 'team_id'] + self._pls_get_safe_fields()

# prepare frequencies dict and handle 'tag_ids'.
# remove tag_ids from fields as it needs specific treatment
use_tags =  bool(fields.pop(fields.index('tag_ids')) if 'tag_ids' in fields else False)
frequencies = dict((field, {}) for field in fields)
if use_tags:
    frequencies = self._pls_update_frequency_table_tag(frequencies, team_id, pls_start_date)
frequencies = self._pls_update_frequency_table_fields(frequencies, stage_ids, stage_sequences, fields, team_id, pls_start_date)
...
```
then inside  _pls_update_frequency_table_tag, do
```py
...
if tag_results:
    # Add 'tag_id' to the frequency dict ('tag_ids' renamed into 'tag_id' to store it properly in frequency table)
    frequencies = {**frequencies, **{'tag_id': {}}}
    for result in tag_results:
        ...
```
this way we don't even add the tag_id dict in 'frequencies' if it's not necessary.
and obviously, don't call _pls_update_frequency_table_tag inside _pls_update_frequency_table_fields anymore. :smile: